pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+
+		if !q.needToBuildAgain(uuid) {
+			return
+		}
+
+		select {
+		case <-q.stopWorker:
+			return
+		default:
+		}
+	}
+}
+
+// PreemptExclusive enqueues a build at the front of the queue and provides
+// exclusive build access. All other builds enqueued via Enqueue()
+// PreemptExclusive() have to finish before the exclusive build is executed.
+// The exclusive build will then be the only build executed until it finishes.
+//
+// If an exclusive build for the same UUID is already enqueued but not yet
+// running, the build will be folded into the already scheduled build but both
+// notification channels will be notified.
+func (q *BuildQueue) PreemptExclusive(b Builder) BuildNotification {
+	notify, enqueue := q.enqueueBuild(b)
+	if enqueue {
+		go q.runExclusiveBuild(b)
+	}
+
+	return notify
+}
+
+// nominateBuild is called when a build has made it to the front of the build
+// queue or the build is a preemptive build. nominateBuild will pull the
+// build information from the buildStatus.nextBuild and nominate the build as
+// buildStatus.currentBuild. nominateBuild() returns the build or nil if the
+// build has since been cancelled.
+func (q *BuildQueue) nominateBuild(uuid string) *build {
+	q.mutex.Lock()
+
+	buildStatus, ok := q.buildStatus[uuid]
+	if !ok || buildStatus.nextBuild == nil {
+		// The builder has been removed since the build was
+		// scheduled. Cancel the build.
+		q.mutex.Unlock()
+		return nil
+	}
+
+	// Mark the scheduled build as building
+	currentBuild := buildStatus.nextBuild
+	buildStatus.currentBuild = currentBuild
+	buildStatus.nextBuild = nil
+	q.mutex.Unlock()
+
+	num := currentBuild.numFolded()
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Sub(float64(num))
+
+	buildStatus.builder.BuildsDequeued(num, false)
+
+	return currentBuild
+}
+
+func (q *BuildQueue) needToBuildAgain(uuid string) bool {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	buildStatus, ok := q.buildStatus[uuid]
+	var nextBuild *build
+	if ok {
+		buildStatus.currentBuild = nil
+		nextBuild = buildStatus.nextBuild
+
+		// If no next build is scheduled, the builder can be
+		// removed from the build status entirely
+		if nextBuild == nil {
+			delete(q.buildStatus, uuid)
+		}
+	}
+
+	return nextBuild != nil
+}
+
+func (q *BuildQueue) runWorker() {
+	for b := range q.workerBuildQueue {
+		uuid := b.GetUUID()
+
+		currentBuild := q.nominateBuild(uuid)
+		if currentBuild == nil {
+			return
+		}
+
+		q.regularBuilds.waitForBuildCondition(q)
+		err := q.build(b)
+		q.regularBuilds.finish(q)
+
+		currentBuild.reportStatus(q, err == nil)
+
+		// If another build for the same builder is scheduled, queue it
+		if q.needToBuildAgain(uuid) {
+			q.workerBuildQueue <- b
+		}
+
+		select {
+		case <-q.stopWorker:
+			return
+		default:
+		}
+	}
+}
+
+// checkBuildGuarantees reports errors when build guarantees are currently not
```

Comment:
guarantees -> invariants (?)

Reply:
I've changed it to build condition to be consistent in naming.